File: src/SignalPhase.cpp
#include "SignalPhase.h"
#include <utility>

SignalPhase::SignalPhase(std::string name, int durationSeconds, std::unordered_map<std::string, LightState> states)
    : name(std::move(name)), durationSeconds(durationSeconds), approachStates(std::move(states)) {}

const std::string &SignalPhase::getName() const { return name; }

int SignalPhase::getDurationSeconds() const { return durationSeconds; }

const std::unordered_map<std::string, LightState> &SignalPhase::getStates() const { return approachStates; }

SignalPhase SignalPhase::allRed(const std::vector<std::string> &approaches) {
    std::unordered_map<std::string, LightState> states;
    for (const auto &approachId : approaches) {
        states[approachId] = LightState::AllRed;
    }
    return SignalPhase("all-red", 5, std::move(states));
}

File: python/cross_roads_ai/signal_phase.py
from typing import Dict
from .light_state import LightState

class SignalPhase:
    def __init__(self, name: str, duration_seconds: float, approach_states: Dict[str, LightState]):
        self._name = name
        self._duration_seconds = duration_seconds
        # approach_states: mapping from approach id -> LightState
        self._approach_states = dict(approach_states)

    def getName(self) -> str:
        return self._name

    def getDurationSeconds(self) -> float:
        return self._duration_seconds

    def approachStates(self) -> Dict[str, LightState]:
        return dict(self._approach_states)
